plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -623,4 +636,111 @@ private static void validateOrcBloomFilterColumns(ConnectorTableMetadata tableMe
+                        }
+                    }
+                }
+            }
+            if (latestMetadataLocation == null) {
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "No metadata file exists at location: " + location);
+            }
+            if (latestMetadataVersionCount > 1) {
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("More than one latest metadata file found at location: %s", location));
+            }
+        }
+        catch (IOException e) {
+            throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed checking table's location: " + location, e);
+        }
+        return Optional.of(latestMetadataLocation);
+    }
+
+    /**
+     * Get the latest metadata file location present in location if metadataFileName is not provided, otherwise
+     *
+     * @param fileSystem TrinoFileSystem
+     * @param location Table location
+     * @param metadataFileName metadata file name
+     * @return full qualified metadata file location
+     */
+    public static Optional<String> getMetadataLocation(TrinoFileSystem fileSystem, String location, String metadataFileName)
+    {
+        if (metadataFileName == null) {
+            return getLatestMetadataLocation(fileSystem, location);
+        }
+        else {
+            return Optional.of(stripTrailingSlash(location) + File.separator + METADATA_FOLDER_NAME + File.separator + metadataFileName);
+        }
+    }
+
+    private static int parseVersion(String metadataLocation)
+    {
+        int versionStart = metadataLocation.lastIndexOf('/') + 1; // if '/' isn't found, this will be 0
+        int versionEnd = metadataLocation.indexOf('-', versionStart);
+        try {
+            return parseInt(metadataLocation.substring(versionStart, versionEnd));
+        }
+        catch (NumberFormatException | IndexOutOfBoundsException e) {
+            return -1;
+        }
+    }
+
+    /**
+     * Read metadata from provided metadata file location. This API should be called for valid metadata location.
+     *
+     * @param fileSystem TrinoFileSystem
+     * @param metadataLocation Metadata file location
+     * @return TableMetadata
+     */
+    public static TableMetadata readMetadata(TrinoFileSystem fileSystem, String metadataLocation)
+            throws TrinoException
+    {
+        FileIO fileIO = fileSystem.toFileIo();
+        InputFile metadataFile = fileIO.newInputFile(metadataLocation);
+        return TableMetadataParser.read(fileIO, metadataFile);
+    }
+
+    /**
+     * Validate provided location
+     *
+     * @param fileSystem TrinoFileSystem
+     * @param location location to validated
+     */
+    public static void validateLocation(TrinoFileSystem fileSystem, String location)
+    {
+        try {
+            if (!fileSystem.newInputFile(location).exists()) {
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("Location %s does not exist", location));
```

Comment:
The error code is not appropriate. The file  system is fine, is just that the file at the location provided is missing.